### PR TITLE
ci: compare bundle size

### PR DESCRIPTION
**Pull request by EdamAme-x**
### The author should do the following, if applicable

- [ ] Add tests
- [x] Run tests
- [x] `bun run format:fix && bun run lint:fix` to format the code
- [ ] Add [TSDoc](https://tsdoc.org/)/[JSDoc](https://jsdoc.app/about-getting-started) to document the code

### CODE REVIEW

**Comment by EdamAme-x** [comment]
if this pr is merged, `before.js` will also be output to show the change in bundle size.

image

**Comment by EdamAme-x** [comment]
Hi @m-shaka, could you please check this?

File: perf-measures/bundle-check/generated/.gitkeep
Comment (yusukebe) [inline perf-measures/bundle-check/generated/.gitkeep:1]: We should add a `perf-measures/bundle-check/.gitignore` like: https://github.com/honojs/hono/blob/main/perf-measures/type-check/.gitignore

**Comment by yusukebe** [comment]
Hi @k2tzumi! Can you review this, you too?

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -223,3 +225,44 @@ jobs:
         with:
           path: perf-measures/type-check/previous-result.txt
           key: type-check-perf-previous-result-${{ github.sha }}
+
+  perf-measures-bundle-check-on-pr:
+    name: 'Bundle Check on PR'
+    runs-on: ubuntu-latest
+    if: github.event_name == 'pull_request'
```
Comment (k2tzumi) [inline .github/workflows/ci.yml:232]: q: Does the bundle size change every time?
Or does it change only when a specific file is changed?

Should I integrate workflow jobs when importing into https://github.com/honojs/hono/pull/3491? Should we keep them separate? I am pondering.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -223,3 +225,44 @@ jobs:
...
+      - uses: actions/cache/restore@v4
+        with:
+          path: perf-measures/bundle-check/generated/before.js
+          restore-keys: |
+            perf-measures-bundle-check-previous-file-
+          key: perf-measures-bundle-check-previous-file-
+      - run: |
+          {
+            echo 'COMPARISON<<EOF'
+            ls -l ./generated | column -s '|' -t
```
Comment (k2tzumi) [inline .github/workflows/ci.yml:248]: It seems that it is not enough to simply parse the result of the ls command to get output like the following
https://github.com/honojs/hono/actions/runs/11818332808/job/32925651835?pr=3665#step:9:24

The `column -s '|' -t` command formats `|` delimited characters, so you need to pipe at least `|` delimited output

ex)
```
| | Current | Previous |
| --- | --- | --- |
| Files | 256 | 256 |
| Lines | 115024 | 115024 |
| Identifiers | 112049 | 112049 |
| Symbols | 243346 | 243346 |
| Types | 204684 | 204684 |
| Instantiations | 3044052 | 3044052 |
| Memory used | 422387K | 422692K |
| I/O read | 0.01s | 0.02s |
| I/O write | 0.00s | 0.00s |
| Parse time | 0.28s | 0.28s |
| Bind time | 0.11s | 0.11s |
| Check time | 2.40s | 2.42s |
| Emit time | 0.00s | 0.00s |
| Total time | 2.79s | 2.81s |
```

File: .github/workflows/ci.yml (hunk 1 above)
Comment (EdamAme-x) [reply .github/workflows/ci.yml:232]: > q: Does the bundle size change every time?
> Or does it change only when a specific file is changed?

`it change only when a specific file is changed` is right.

Only when a core file is changed.

> Should I integrate workflow jobs when importing into https://github.com/honojs/hono/pull/3491? Should we keep them separate? I am pondering.  

I have limited knowledge about it, sorry.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (EdamAme-x) [reply .github/workflows/ci.yml:248]: thanks, any idea to format the output?

Comment (EdamAme-x) [reply .github/workflows/ci.yml:248]: thanks, any idea to format the output?

Comment (k2tzumi) [reply .github/workflows/ci.yml:248]: How about the following format?

| | Current | Previous |
| --- | --- | --- |
|  Bundle size | 19206 | 19206 |

The number of bytes assumes that the number of digits will not change in the future.

Comment (yusukebe) [reply .github/workflows/ci.yml:248]: Looks good to me!

**Comment by EdamAme-x** [comment]
what happen, ddos...?
image

File: .github/workflows/ci.yml (hunk 1 above)
Comment (EdamAme-x) [reply .github/workflows/ci.yml:232]: I would like to keep this as a TODO/NOTE as it seems a bit difficult to implement.

https://stackoverflow.com/questions/70708306/github-actions-run-step-job-in-a-workflow-if-changes-happen-in-specific-folde

File: .github/workflows/ci.yml (hunk 2 above)
Comment (EdamAme-x) [reply .github/workflows/ci.yml]: image
@k2tzumi what do you think?

Comment (EdamAme-x) [reply .github/workflows/ci.yml]: It is simple, but it is good enough for now.
Let's add custom metrics later.

**Comment by EdamAme-x** [comment]
hi @yusukebe, this pr should be ready.

**Review by k2tzumi — APPROVED** [review]
Looks good to me!

**Review by m-shaka — APPROVED** [review]
Sorry for being late. Great work!

It would be nicer if you could see a unit(like KB or MB) of bundle size, but it's OK to add it later

**Comment by EdamAme-x** [comment]
> It would be nicer if you could see a unit(like KB or MB) of bundle size, but it's OK to add it later

Agreed!  
However, we will add custom metrics later, so it seems like a good idea to add them at that time.

**Review by yusukebe — APPROVED** [review]
LGTM!

**Comment by yusukebe** [comment]
@EdamAme-x @k2tzumi @m-shaka

Thank you all! Let's go with this.

**Pull request merged**